File: src - c++/Esercizi/ContaNodiNonFoglia/ContaNodiNonFoglia.cpp
//Metodo privato
unsigned int AlberoBinario::_ContaNodiNonFoglia(const PNodo& n) const {
  if (!n)
    return 0;
    
  unsigned int count = 0;
    
  //eventuale contributo sottoalbero sinistro
  if (n->sin)
    count = count + _ContaNodiNonFoglia(n->sin);

  //eventuale contributo sottoalbero destro
  if (n->des)
    count = count + _ContaNodiNonFoglia(n->des);
    
  //eventuale contributo del presente nodo
  if (n->sin || n->des)
    count++;
    
  return count;    
}
  
//Metodo pubblico
unsigned int AlberoBinario::ContaNodiNonFoglia() const {
  return _ContaNodiNonFoglia(root);  
}

File: src - java/Esercizi/ContaNodiNonFoglia/ContaNodiNonFoglia.java
	private boolean eUnaFoglia(Nodo n) {
		// metodo di supporto che verifica se il nodo
		// puntato da n è o meno una foglia.
		return (n.sin == null) && (n.des == null);
	}
	
	private int _contaNodiNonFoglia(Nodo n) {
		if (n == null) // Se l'albero è vuoto
			return 0; // il conteggio è nullo

		int count = 0; // accumulatore

		// eventuale contributo sottoalbero sinistro
		if (n.sin != null)
			count = count + _contaNodiNonFoglia(n.sin);

		// eventuale contributo sottoalbero destro
		if (n.des != null)
			count = count + _contaNodiNonFoglia(n.des);

		// eventuale contributo del presente nodo
		if (!eUnaFoglia(n))
			count++;

		return count;
	}

	public int contaNodiNonFoglia() {
		return _contaNodiNonFoglia(root);
	}